node/router/src/lib.rs

```diff
@@ -183,6 +199,79 @@ pub trait PeerPoolHandling<N: Network>: P2P {
         }
     }
 
+    /// Downgrades a connected peer to candidate status.
+    fn downgrade_peer_to_candidate(&self, listener_addr: SocketAddr) {
+        if let Some(peer) = self.peer_pool().write().get_mut(&listener_addr) {
+            if let Peer::Connected(peer) = peer {
+                self.resolver().write().remove_peer(peer.connected_addr, peer.aleo_addr);
+            }
+            peer.downgrade_to_candidate(listener_addr);
+        }
+    }
+
+    /// Adds new candidate peers to the peer pool, ensuring their validity and following the
+    /// limit on the number of peers in the pool.
+    fn insert_candidate_peers(&self, mut listener_addrs: Vec<SocketAddr>) {
+        // Perform filtering to ensure candidate validity.
+        {
+            let peer_pool = self.peer_pool().read();
+            listener_addrs.retain(|&addr| {
+                !peer_pool.contains_key(&addr)
+                    && !self.is_ip_banned(addr.ip())
+                    && if self.is_dev() { !is_bogon_ip(addr.ip()) } else { self.is_valid_peer_ip(addr) }
+            });
+        }
+
+        // If we've managed to filter out every entry, there's nothing to do.
+        if listener_addrs.is_empty() {
+            return;
+        }
+
+        // If we're about to exceed the peer pool size limit, apply candidate slashing.
+        if self.number_of_peers() + listener_addrs.len() >= Self::MAXIMUM_POOL_SIZE && Self::PEER_SLASHING_COUNT != 0 {
```

Comment:
If we're close to the limit, and we receive multiple PeerResponses, might we trigger this logic multiple times, potentially slashing down to 0?

Granted, the snapshot is unlikely to be very different, so we'll likely slash the same peers.

Holding a lock should not be too much of an issue because this function is just a background utility. In fact, doing the slashing serially might be more kind to CPU.